### ISSUE TITLE
Add a mechanism to override how Blazor's JS dynamic imports are performed

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/aspnetcore#45148]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

When you do `JS.InvokeAsync("import", "./some/module.js")`, the JS interop system will perform a JS dynamic import for `"./some/module.js"`. We even recommend that package authors use this technique to load their JS dependencies dynamically so the package consumer doesn't have to add any `<script>` tags manually.

However, in some deployment scenarios, the app author wants to override how *all* static resources are loaded. The app author can control Blazor boot resource loading (e.g., .NET assemblies, our baked-in JS library, and the `.wasm` file) using a `loadBootResource` callback. But they have no comparable way to override the URL used by a JSInterop-triggered `import`.

Originally reported at https://github.com/dotnet/aspnetcore/issues/45104

### Describe the solution you'd like

We could consider either:

 1. A general-purpose way to register a callback that handles JS dynamic import
 2. Making JS dynamic import call through the existing `loadBootResource` API (which is admittedly strange since those files are loaded arbitrarily when the app is running, not as part of its boot process - they aren't boot resources as per normal definition)

### Additional context

A counterargument would be that JS dynamic imports are only one of many other static file types that may be loaded by a component in a package. They might also dynamically load images, CSS, JSON files via `fetch`, or whatever else. In all those cases you wouldn't have a way to intercept and modify their URLs, so it would be strange to special-case a feature for JS imports.

We could argue that a better, more all-encompassing solution is for such developers to implement a service worker that gives them full control over all the static resources fetched by their site.

**Comment by javiercn** [comment]
> However, in some deployment scenarios, the app author wants to override how _all_ static resources are loaded. The app author can control Blazor boot resource loading (e.g., .NET assemblies, our baked-in JS library, and the `.wasm` file) using a `loadBootResource` callback. But they have no comparable way to override the URL used by a JSInterop-triggered `import`.

Yep, we need something for JsInitializers too, since this has been brought up a couple of times.

**Comment by javiercn** [comment]
> 2\. Making JS dynamic import call through the existing `loadBootResource` API (which is admittedly strange since those files are loaded arbitrarily when the app is running, not as part of its boot process - they aren't boot resources as per normal definition)

This won't work for Server scenarios though :(

**Comment by javiercn** [comment]
> A counterargument would be that JS dynamic imports are only one of many other static file types that may be loaded by a component in a package. They might also dynamically load images, CSS, JSON files via `fetch`, or whatever else. In all those cases you wouldn't have a way to intercept and modify their URLs, so it would be strange to special-case a feature for JS imports.

I think we already "interfere" with the original request, so I think it is fair we give people control back.

**Comment by javiercn** [comment]
> We could argue that a better, more all-encompassing solution is for such developers to implement a service worker that gives them full control over all the static resources fetched by their site.

Maybe, but how would that work on first load?

**Comment by martinlingstuyl** [comment]
It's a bit of an edge case maybe (or is it?), but what if you _can_ control static clientside resources, but the server itself is out of reach? A service worker/proxy is no solution in that case.

**Comment by martinlingstuyl** [comment]
What I mean is that I _can_ control the CDN or server that hosts the WASM files AND I can control static files on a website where the Blazor WASM app needs to be incorporated, but I don't control the actual website or server, so I cannot configure a proxy.

**Comment by SteveSandersonMS** [comment]
> Maybe, but how would that work on first load?

Wait until the service worker is loaded before calling `Blazor.start`? Not sure if this necessitates auto-refreshing the page.

> what if you can control static clientside resources, but the server itself is out of reach? A service worker/proxy is no solution in that case.

Service workers are just static content loaded from the CDN/server.

**Comment by martinlingstuyl** [comment]
> Service workers are just static content loaded from the CDN/server.

In that case, I think I need to do some reading up!

**Comment by cveld** [comment]
I believe I am currently blocked with my use case:
* vscode extension webview that hosts a blazor wasm app
* I am trying to inject specific configuration before blazor main is started so that I can implement differentiated logic compared to the regular blazor wasm web app experience
* using the onRuntimeReady javascript initializer
* overriding the location of the boot static assets with the `loadBootResource` hook

`dotnet.js` currently tries to load my module `myapp.lib.module.js` but does that from the wrong path:

```
GET vscode-webview://0knolpkv5olkqn3b8frsm6kntt3sfqn3snrbojmgmbn9ucck8h2c/myapp.lib.module.js net::ERR_ACCESS_DENIED
```

vscode webview api provides `asWebviewUri` to get the correct uri. But it seems that Blazor currently does not provide means to override the base path for these dynamic imports. `asWebviewUri` generates https://file+.vscode-resource.vscode-cdn.net based uris.

**Comment by KristofferStrube** [comment]
@javiercn, It could be cool if this could be done by Copilot. I have many that have requested a feature like this in my File System and File System Access libraries, as these are often used for making PWAs, browser extensions, and standalone Electron apps that have constraints on what folders they can use. That has led me to implement some workarounds that I'm not satisfied with, as I also believe it should be controlled centrally by the framework itself.

My favourite idea would be something close to what Steve Sanderson recommended:
> A general-purpose way to register a callback that handles JS dynamic import

It could be something like:
```csharp
WebAssemblyHost host = builder.Build();

host.UseJSInteropImportRouting(path =>
{
    const string startToChange = "./_content/";
    return path?.StartsWith(startToChange) is true ? $"./{path[3..]}" : path;
});
```

Alternative names could be:
- MapJSInteropImportPath
- UseJSInteropImportPath
- MapAlternativeJSInteropImportPath
- UseAlternativeJSInteropImportPath
- MapCustomJSInteropImportPath
- UseCustomJSInteropImportPath

**Comment by javiercn** [comment]
@KristofferStrube actually, reading this again, I don't think we should do this.

We would not offer a c# api for this, because it's not a c# concern, we would at best offer a JS API. 

However, this is precisely what `<script type="importmap"></script>` is for, which you can do passing a custom importmap to the `ImportMap` component.

I do not think we need or should add a secondary mechanism for this.

@KristofferStrube if you want, give the suggested approach a try, and if for some reason you aren't able to make it work, file a separate issue and we will reconsider.

**Comment by KristofferStrube** [comment]
@javiercn, I will try it out. 😃 Good idea.
